packages/block-editor/src/components/provider/index.js

```diff
@@ -12,6 +12,10 @@ import { compose } from '@wordpress/compose';
 import withRegistryProvider from './with-registry-provider';
 
 class BlockEditorProvider extends Component {
+	constructor() {
+		super( ...arguments );
+		this.lastPersistedBlocks = [];
```

Comment:
I'm not satisfied with this trick (this value being an array) but it's the only way I found right now to fix all the use-cases which are:

 - We need to avoid calling `resetBlocks` if the new "value" prop received is the same as the one already in the block-editor store (to avoid resetting the "undo" levels). (undo e2e test failing)
 - Sometimes calling some changes like inserting a block with a template (columns block for instance) result in two `onChange` calls being performed synchronously which will eventually trigger two `componentDidUpdate` calls but eventually the first `componentDidUpdate` will call `resetBlocks` if we only compare against the latest blocks value in the block-editor state causing the editor to reset to the previous blocks value (and not the last one). (columns e2e test failing)

The fix here is a bit fragile though, it basically says: don't call "resetBlocks" unless the new prop received is not  one of the last persisted blocks value passed to `onChange`.